File: sort/selection_sort/kotlin/selection_sort.kt
fun <T:Comparable<T>>selectionsort(items:MutableList<T>):MutableList<T>{
    if (items.isEmpty()){
        return items
    }
    for (idx in 0..items.count()){
        val array = items.subList(0,items.count()-idx)
        val minItem = array.min()
        val indexOfMinItem = array.indexOf(minItem)

        if (minItem != null) {
            items.removeAt(indexOfMinItem)
            items.add(minItem)
        }
    }
    return items
}


fun main(args: Array<String>) {
    println("Selection Sort")
    val names = mutableListOf("John", "Tim", "Zack", "Daniel", "Adam")
    println(names)
    var ordered = insertionsort(names)
    println(ordered)
}

File: data_structures/Arrays/ruby/selection_sort.rb
def selection_sort(array)
  return array if array.size <= 1

  0.upto(array.length-2) do |i|
    min = i
    (i+1).upto(array.length-1) { |k| min = k if array[k] < array[min] }
    array[i], array[min] = array[min], array[i] if i != min
  end

  return array
end
